File: api-tester/src/defaults/java/com/revenuecat/apitester/java/revenuecatui/FontProviderAPI.java
package com.revenuecat.apitester.java.revenuecatui;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.OptIn;
import androidx.compose.ui.text.font.FontFamily;
import androidx.compose.ui.text.font.FontWeight;
import androidx.compose.ui.text.googlefonts.GoogleFont;
import com.revenuecat.purchases.ui.revenuecatui.fonts.CustomFontProvider;
import com.revenuecat.purchases.ui.revenuecatui.fonts.CustomParcelizableFontProvider;
import com.revenuecat.purchases.ui.revenuecatui.fonts.FontProvider;
import com.revenuecat.purchases.ui.revenuecatui.fonts.GoogleFontProvider;
import com.revenuecat.purchases.ui.revenuecatui.fonts.ParcelizableFontProvider;
import com.revenuecat.purchases.ui.revenuecatui.fonts.PaywallFont;
import com.revenuecat.purchases.ui.revenuecatui.fonts.PaywallFontFamily;
import com.revenuecat.purchases.ui.revenuecatui.fonts.TypographyType;

import java.util.List;

@SuppressWarnings({"unused"})
final class FontProviderAPI {
    static void check(TypographyType typographyType) {
        FontProvider fontProvider = new FontProvider() {
            @Nullable
            @Override
            public FontFamily getFont(@NonNull TypographyType type) {
                return null;
            }
        };
        FontFamily fontFamily = fontProvider.getFont(typographyType);
        CustomFontProvider customFontProvider = new CustomFontProvider(FontFamily.Companion.getDefault());
    }

    static void checkParcelizableFontProvider(TypographyType typographyType, List<PaywallFont> fonts) {
        ParcelizableFontProvider fontProvider = new ParcelizableFontProvider() {
            @Nullable
            @Override
            public PaywallFontFamily getFont(@NonNull TypographyType type) {
                return null;
            }
        };
        PaywallFontFamily fontFamily = fontProvider.getFont(typographyType);
        CustomParcelizableFontProvider customFontProvider = new CustomParcelizableFontProvider(new PaywallFontFamily(fonts));
    }

    static void checkPaywallFontFamily(PaywallFontFamily fontFamily) {
        List<PaywallFont> fonts = fontFamily.getFonts();
    }

    static void checkPaywallFont(PaywallFont font) {
        if (font instanceof PaywallFont.GoogleFont) {
            final PaywallFont.GoogleFont googleFont = (PaywallFont.GoogleFont) font;
            String fontName = googleFont.getFontName();
            GoogleFontProvider provider = googleFont.getFontProvider();
            FontWeight fontWeight = googleFont.getFontWeight();
            int fontStyle = googleFont.getFontStyle();
            final PaywallFont.GoogleFont googleFont2 = new PaywallFont.GoogleFont(
                    fontName,
                    provider,
                    fontWeight,
                    fontStyle
            );
        } else if (font instanceof PaywallFont.ResourceFont) {
            final PaywallFont.ResourceFont resourceFont = (PaywallFont.ResourceFont) font;
            int fontRes = resourceFont.getResourceId();
            FontWeight fontWeight = resourceFont.getFontWeight();
            int fontStyle = resourceFont.getFontStyle();
            final PaywallFont.ResourceFont resourceFont2 = new PaywallFont.ResourceFont(fontRes, fontWeight, fontStyle);
        }
    }

    static void checkGoogleFontProvider(
            int certificates,
            String providerAuthority,
            String providerPackage
    ) {
        GoogleFontProvider provider = new GoogleFontProvider(certificates, providerAuthority, providerPackage);
        int providerCertificates = provider.getCertificates();
        String providerAuthority2 = provider.getProviderAuthority();
        String providerPackage2 = provider.getProviderPackage();
        GoogleFont.Provider googleProvider = provider.toGoogleProvider();
    }

    static Boolean checkTypographyType(TypographyType type) {
        switch (type) {
            case DISPLAY_LARGE:
            case DISPLAY_MEDIUM:
            case DISPLAY_SMALL:
            case HEADLINE_LARGE:
            case HEADLINE_MEDIUM:
            case HEADLINE_SMALL:
            case TITLE_LARGE:
            case TITLE_MEDIUM:
            case TITLE_SMALL:
            case BODY_LARGE:
            case BODY_MEDIUM:
            case BODY_SMALL:
            case LABEL_LARGE:
            case LABEL_MEDIUM:
            case LABEL_SMALL:
                return true;
        }
        return false;
    }
}

File: api-tester/src/defaults/kotlin/com/revenuecat/apitester/kotlin/revenuecatui/FontProviderAPI.kt
package com.revenuecat.apitester.kotlin.revenuecatui

import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontFamily.Companion.Default
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.googlefonts.GoogleFont
import com.revenuecat.purchases.ui.revenuecatui.ExperimentalPreviewRevenueCatUIPurchasesAPI
import com.revenuecat.purchases.ui.revenuecatui.fonts.CustomFontProvider
import com.revenuecat.purchases.ui.revenuecatui.fonts.CustomParcelizableFontProvider
import com.revenuecat.purchases.ui.revenuecatui.fonts.FontProvider
import com.revenuecat.purchases.ui.revenuecatui.fonts.GoogleFontProvider
import com.revenuecat.purchases.ui.revenuecatui.fonts.ParcelizableFontProvider
import com.revenuecat.purchases.ui.revenuecatui.fonts.PaywallFont
import com.revenuecat.purchases.ui.revenuecatui.fonts.PaywallFontFamily
import com.revenuecat.purchases.ui.revenuecatui.fonts.TypographyType

@Suppress("unused", "UNUSED_VARIABLE")
@OptIn(ExperimentalPreviewRevenueCatUIPurchasesAPI::class)
private class FontProviderAPI {
    fun check(typographyType: TypographyType) {
        val fontProvider: FontProvider = object : FontProvider {
            override fun getFont(type: TypographyType): FontFamily? {
                return null
            }
        }
        val fontFamily = fontProvider.getFont(typographyType)
        val customFontProvider = CustomFontProvider(Default)
    }

    fun checkParcelizableFontProvider(typographyType: TypographyType, fonts: List<PaywallFont>) {
        val fontProvider: ParcelizableFontProvider = object : ParcelizableFontProvider {
            override fun getFont(type: TypographyType): PaywallFontFamily? {
                return null
            }
        }
        val fontFamily = fontProvider.getFont(typographyType)
        val customFontProvider = CustomParcelizableFontProvider(PaywallFontFamily(fonts))
    }

    fun checkPaywallFontFamily(fontFamily: PaywallFontFamily) {
        val fonts = fontFamily.fonts
    }

    fun checkPaywallFont(font: PaywallFont) {
        if (font is PaywallFont.GoogleFont) {
            val fontName: String = font.fontName
            val provider: GoogleFontProvider = font.fontProvider
            val fontWeight: FontWeight = font.fontWeight
            val fontStyle: Int = font.fontStyle
            val googleFont = PaywallFont.GoogleFont(
                fontName,
                provider,
                fontWeight,
                fontStyle,
            )
        } else if (font is PaywallFont.ResourceFont) {
            val fontRes: Int = font.resourceId
            val fontWeight: FontWeight = font.fontWeight
            val fontStyle: Int = font.fontStyle
            val resourceFont = PaywallFont.ResourceFont(fontRes, fontWeight, fontStyle)
        }
    }

    fun checkGoogleFontProvider(
        certificates: Int,
        providerAuthority: String?,
        providerPackage: String?,
    ) {
        val provider = GoogleFontProvider(certificates, providerAuthority!!, providerPackage!!)
        val providerCertificates = provider.certificates
        val providerAuthority2 = provider.providerAuthority
        val providerPackage2 = provider.providerPackage
        val googleProvider: GoogleFont.Provider = provider.toGoogleProvider()
    }

    fun checkTypographyType(type: TypographyType): Boolean {
        when (type) {
            TypographyType.DISPLAY_LARGE,
            TypographyType.DISPLAY_MEDIUM,
            TypographyType.DISPLAY_SMALL,
            TypographyType.HEADLINE_LARGE,
            TypographyType.HEADLINE_MEDIUM,
            TypographyType.HEADLINE_SMALL,
            TypographyType.TITLE_LARGE,
            TypographyType.TITLE_MEDIUM,
            TypographyType.TITLE_SMALL,
            TypographyType.BODY_LARGE,
            TypographyType.BODY_MEDIUM,
            TypographyType.BODY_SMALL,
            TypographyType.LABEL_LARGE,
            TypographyType.LABEL_MEDIUM,
            TypographyType.LABEL_SMALL,
            -> return true
        }
    }
}
